### ISSUE TITLE
Remove duplicates on large number of files

### ISSUE THREAD

**Issue by im7mortal** [ente-io/ente#568]
I am trying to sync my google zip file with ente. I tried it to 2 years back #ente-io/photos-web/issues/243 and had problems. 
So turn out that it every time created new file. Now I have 2 or 3 duplicates for at least 15 000 files. 
![ENTE_SELECTION](https://github.com/ente-io/photos-desktop/assets/5336231/60e50781-9e8e-4d16-9fbe-e481e7e36632)

1. I just don't want to do it again. It's exhausting and interface doesn't help.
2. To revise it manually it's huge pain. I was trying to do it and after 1 hour the app was down for memory exceed. 
3. I can't trust ente as you do comparison of duplicates by size. And I saw many duplicates with important to me fotos. 

So how I can see the possible solution

1. Maybe limit duplicates by chunks. For example 1000 duplicates. Then it possible to remove this chunk and start new one.
2. Memory crash should be debugged
3. Add sophisticated filter to the duplicate search. : name, size, I believe you should have a some kind of hash also? 

Some more prehistory . I was trying to migrate to ente in 2021 but I couldn't because of the bugs. Then I had my phone connected to ente. So I just can't remove all files and start google zip import from the scratch.

**Comment by im7mortal** [comment]
Hi @vishnukvmd !

I didn't try to clean duplicates from time we discussed it.


I saw there is AI feature now which will handle images locally. So I am thinking why do not also generate hashes for all images in the same run?

**Comment by mnvr** [comment]
Yes. No ETA, but we've sketched out some approaches for using form of perceptual hashing or cosine similarity based deduping, and hope to get around to working on it at some point soon.

**Comment by im7mortal** [comment]
@mnvr 

What I understand from [this comment](https://github.com/ente-io/ente/issues/568#issuecomment-1975055699). Ente create simple hash on first upload of the image. So I was advised to erase all photos and reupload them to generate hash for all old photos. 

![111](https://github.com/user-attachments/assets/b0db2dcf-860c-4a78-b3ce-7dde00302565)

Why don't generate the hash during indexing and do not update metadata? It will be quick effective solution. 


I mean if we compare with perceptual hashing or cosine similarity which are more sophisticated , the  hash generation is extremely straight forward and will not take time to implement it.

**Comment by im7mortal** [comment]
@vishnukvmd @mnvr @abhinavkgrd  I am very upset about [this comment](https://github.com/ente-io/ente/blob/main/web/packages/new/photos/services/dedup.ts#L136-L141).

```golang
        if (!hash) {
            // Some very old files uploaded by ancient versions of Ente might
            // not have hashes. Ignore these.
            continue;
        }

```


These files without a hash are from the earliest  Ente supporters.
I also believed in Ente and I uploaded all my Google Photos to Ente.
I did it a couple of times because of **bugs** in the _ancient versions of Ente_.
Now I’ve learned that you’ve decided to _ignore_ all my memories.

**Comment by vishnukvmd** [comment]
Hi @im7mortal, sorry you feel that way. We aren't ignoring all your memories, the items where hashes are unavailable are being ignored from the de-duplication flow.

We had considered backfilling data for these, but had taken a step back to avoid complexity that will have to be maintained in perpetuity.

Also, we intend to provide a CLIP embeddings based de-duplication flow in the future, that will detect similar images, in addition to exact duplicates — that will work for all customers.

All this said, we understand the troubles, and if you'd like, support@ente.io would be happy to help set up a fresh account, issuing a refund for your existing subscription.

**Comment by im7mortal** [comment]
@vishnukvmd Thank you for quick response! :smiley:


> We had considered backfilling data for these, but had taken a step back to avoid complexity that will have to be maintained in perpetuity.

1. Could I please understand what the problem is with backfilling in this scenario? Can’t we just update the hashes in the metadata? Could we maybe do it on demand—like when a user triggers deduplication? I’m not sure how many users are affected, but I consider this very important. As I mentioned, I have almost all my memories from the past 10 years unhashed.

>Also, we intend to provide a CLIP embeddings based de-duplication flow in the future, that will detect similar images, in addition to exact duplicates — that will work for all customers.

2. CLIP sounds great, but I still believe it should be built on top of basic hashing.

I created a pull request with simple dynamic hash generation. It's just a proof of concept. It covers half of my needs—I mean, I can delete duplicates with it. The problem is that I still haven’t finished importing from Google Photos (ente-io/photos-web/issues/243 — I haven’t closed it, by the way). So, if I try to import again and clean duplicates, I’ll still have 30GB of unhashed files. Almost all the files from my Google Photos backup are already in Ente—I just need to find the fraction that failed before. I still need a way to filter those out.

**Comment by vishnukvmd** [comment]
Code is baggage, and we'd rather not add code that does not help majority of our customers, but has to be maintained forever.

If you haven't yet completed your import, I'd recommend starting afresh, with a full refund from [support@ente.io](mailto:support@ente.io).

**Comment by im7mortal** [comment]
> If you haven't yet completed your import, I'd recommend starting afresh, with a full refund from [support@ente.io](mailto:support@ente.io).

I’ve mentioned a few times that I used Ente with regular sync from my phone before hashes were introduced, so my library ended up being a mix. It’s really difficult now to identify and filter out another 20 GB of photos that aren’t in my Google Photos backup.

> Code is baggage, and we'd rather not add code that does not help majority of our customers, but has to be maintained forever.

I understand that adding code comes with long-term maintenance costs, and it makes sense to prioritize features that help most users. Still, since hashing was introduced later, there might be more users affected than it appears — and many of them probably don’t even realize it. Did you estimate the number of active users and how many terabytes of images were stored before hashes were introduced? **Would it be possible to publish these numbers please?**


_I just wanted to share my experience._ The original deduplication method was based on name and size — when I saw it, I immediately recognized the threat, and I spent over an hour reviewing thousands of duplicates to make sure my images were safe, and instead proved that the implementation itself was dangerous.

**Comment by im7mortal** [comment]
Hi @vishnukvmd 

Now, with time, I better understand your response, which at first I considered a bit rude. (Of course, I understood that any new feature brings extra workload for developers, but as someone whose data was corrupted due to an earlier misimplementation, I felt affected.)

About a month after our conversation, you released Ente 1.0.0 with stabilized features, and I realized you were under pressure at that time.

I haven’t done anything since then, and I still can’t use Ente normally. I’ll return to [issue 243](https://github.com/ente-io/photos-web/issues/243) in the near future, and then I’ll try to resolve the duplication problem again.
